File: Medium/0230-Kth Smallest Element in a BST/Inorder Traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} k
# @return {Integer}
def kth_smallest(root, k)
    stack = []

    while true
        until root.nil?
            stack.push(root)
            root = root.left
        end

        root = stack.pop
        k -= 1

        return root.val if k == 0

        root = root.right
    end
end

File: Problemset/1038-Binary Search Tree to Greater Sum Tree/Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        stack = []
        curr = root
        sum = 0

        while stack or curr:
            while curr:
                stack.append(curr)
                curr = curr.right

            curr = stack.pop()
            sum += curr.val
            curr.val = sum

            curr = curr.left

        return root
